### PR TITLE
[GraphOptimizer]Eliminate reshape node that follows a splat node.

### PR DESCRIPTION
**Pull request by ZchiPitt**
Modified the GraphOptimizer::optimizeReshape() function, making it able to eliminate reshape node that follows a splat node, by replace with a new splat node.

### CODE REVIEW

File: lib/Optimizer/GraphOptimizer.cpp [hunk 1]
```diff
@@ -1076,14 +1076,22 @@ static void optimizeSliceOfSplat(Function *F) {
...
     auto *reshapeNode = dyn_cast<ReshapeNode>(&node);
     if (!reshapeNode)
       continue;
-    auto inputNode = reshapeNode->getInput();
+    auto &inputNode = reshapeNode->getNthInput(0);
     if (inputNode.dims() == reshapeNode->dims()) {
       reshapeNode->getResult().replaceAllUsesOfWith(inputNode);
+      continue;
+    }
+    auto *splatNode = dyn_cast<SplatNode>(inputNode);
```
Comment (opti-mix) [inline lib/Optimizer/GraphOptimizer.cpp]: Add a comment before this line to describe the transformation, e.g.:
```c++
// Reshape(Splat(args)) -> Splat(args')
```

File: lib/Optimizer/GraphOptimizer.cpp [hunk 2]
```diff
@@ -1076,14 +1076,22 @@ static void optimizeSliceOfSplat(Function *F) {
...
+    if (splatNode && splatNode->hasOneUse()) {
```
Comment (opti-mix) [inline lib/Optimizer/GraphOptimizer.cpp]: Add a comment to explain why only splat nodes with a single used are transformed.

**Comment by opti-mix** [comment]
The name of the component is usually added to the title of the commit message like this:
`[graph-optimizer] Replace reshapes of splat nodes by splats with required shapes`
No dot at the end of the title.

And then the rest of the commit messages describes the commit itself.

**Comment by opti-mix** [comment]
Sorry, clicked on the wrong button ;-)

File: tests/unittests/graphOptzTest.cpp [hunk 3]
```diff
@@ -604,6 +604,46 @@ TEST_F(GraphOptz, ReshapeNoop) {
   EXPECT_TRUE(SN->getType()->dims().equals(shape));
 }
 
+/// Test on the optimation on the case of reshapeNode following splatNode
```
Comment (opti-mix) [inline tests/unittests/graphOptzTest.cpp]: optimation -> optimization

May be replace the line by a more descriptive:
```c++
/// Test the Reshape(Splat(args)) -> Splat(args') transformation
```

File: tests/unittests/graphOptzTest.cpp [hunk 4]
```diff
@@ -604,6 +604,46 @@ TEST_F(GraphOptz, ReshapeNoop) {
...
+TEST_F(GraphOptz, ReshapeAfterSplat) {
```
Comment (opti-mix) [inline tests/unittests/graphOptzTest.cpp]: May be add a comment that you try to test a positive and a negative test-cases? In one case the optimization should kick-in because this and that, and in the other case the optimization should not happen because of this and that.

File: lib/Optimizer/GraphOptimizer.cpp [hunk 5]
```diff
@@ -1076,14 +1076,22 @@ static void optimizeSliceOfSplat(Function *F) {
...
+      auto *newSplatNode = F->createSplat(
+          splatNode->getName(), reshapeNode->getType(), splatNode->getValue());
+      reshapeNode->getResult().replaceAllUsesOfWith(newSplatNode);
```
Comment (jfix71) [inline lib/Optimizer/GraphOptimizer.cpp]: nit: I would prefer to add a `continue;` here to keep with the pattern of peephole loops that check different conditions. Would prevent a potential bug if someone adds a new if case below this one and does not realize there is no continue here.

File: lib/Optimizer/GraphOptimizer.cpp (hunk 1 above)
Comment (ZchiPitt) [reply lib/Optimizer/GraphOptimizer.cpp]: Done.

File: lib/Optimizer/GraphOptimizer.cpp (hunk 2 above)
Comment (ZchiPitt) [reply lib/Optimizer/GraphOptimizer.cpp]: Done.

File: tests/unittests/graphOptzTest.cpp [hunk 6]
```diff
@@ -604,6 +604,46 @@ TEST_F(GraphOptz, ReshapeNoop) {
...
+
+  // Before optimization, we have 9 nodes in the graph.
+  EXPECT_EQ(F_->getNodes().size(), 9);
+
+  ::glow::optimize(F_, CompilationMode::Infer);
+
+  // After optimization, we expect to see only 8 nodes, as Z2,R2 would be
+  // replace by a new splat node.
+  EXPECT_EQ(F_->getNodes().size(), 8);
+
```
[the remark this answers is not in the document]
Comment (ZchiPitt) [reply tests/unittests/graphOptzTest.cpp]: Done.

File: tests/unittests/graphOptzTest.cpp (hunk 3 above)
Comment (ZchiPitt) [reply tests/unittests/graphOptzTest.cpp]: Done.

File: tests/unittests/graphOptzTest.cpp (hunk 4 above)
Comment (ZchiPitt) [reply tests/unittests/graphOptzTest.cpp]: Done.

File: lib/Optimizer/GraphOptimizer.cpp (hunk 5 above)
Comment (ZchiPitt) [reply lib/Optimizer/GraphOptimizer.cpp]: Done.

File: lib/Optimizer/GraphOptimizer.cpp [hunk 7]
```diff
@@ -1076,14 +1076,26 @@ static void optimizeSliceOfSplat(Function *F) {
...
+    auto &inputNode = reshapeNode->getNthInput(0);
     if (inputNode.dims() == reshapeNode->dims()) {
       reshapeNode->getResult().replaceAllUsesOfWith(inputNode);
+      continue;
+    }
+    // Reshape(Splat(args)) -> Splat(args').
+    auto *splatNode = dyn_cast<SplatNode>(inputNode);
+    if (splatNode && splatNode->hasOneUse()) {
+      // Splat node with more than one use can not be transformed, otherwise it
+      // would cause type inconsistency.
```
Comment (opti-mix) [inline lib/Optimizer/GraphOptimizer.cpp]: This is wrong. It has nothing to do with any type inconsistency, as we create a new splat and do not touch the old one. So, types will be consistent.

The reason is that if we do it for all reshape(splats), we may increase the number of splats, which may lead to increased memory consumption during the execution of the NN model.

File: lib/Optimizer/GraphOptimizer.cpp [hunk 8]
```diff
@@ -1076,14 +1076,26 @@ static void optimizeSliceOfSplat(Function *F) {
 }
 
 /// Eliminate ReshapeNode when the input is already the correct shape.
+/// Also eliminate ReshapeNode when the input is SplatNode.
```
Comment (opti-mix) [inline lib/Optimizer/GraphOptimizer.cpp]: I'd suggest moving the `  /// Eliminate ReshapeNode when the input is already the correct shape. ` comment into the body of the function, right before the corresponding `if`. And removing the `/// Also eliminate ReshapeNode when the input is SplatNode` here.

We may add more reshape optimizations in the future. But it doesn't make a lot of sense to list them all in the function's comment.

IMHO, the function's comment should simply say something like:
`/// Optimize reshape nodes.`

File: tests/unittests/graphOptzTest.cpp (hunk 6 above)
[the remark this answers is not in the document]
Comment (opti-mix) [reply tests/unittests/graphOptzTest.cpp]: It is not quite done. You do not check that R1 is literally still in the graph, i.e. as a node. You check that there is a node that looks like R1 ;-)

You also do not check that R3 and Z2 are literally not in the graph. Instead you check the number of nodes in the graph.

[the remark this answers is not in the document]
Comment (ZchiPitt) [reply tests/unittests/graphOptzTest.cpp]: Thx for the suggestions. I use the helper function IsSameNodeAddress() right now to check if R1 still remains and if R3,Z2 got removed.

File: lib/Optimizer/GraphOptimizer.cpp (hunk 8 above)
Comment (ZchiPitt) [reply lib/Optimizer/GraphOptimizer.cpp]: Adopted.

File: lib/Optimizer/GraphOptimizer.cpp (hunk 7 above)
Comment (ZchiPitt) [reply lib/Optimizer/GraphOptimizer.cpp]: Thx for the correction. Changed to the correct comment.
